tests/template_tests/test_library.py

```diff
@@ -0,0 +1,167 @@
+from django.template.base import Node
+from django.test import TestCase
+
+
+class FilterRegistrationTests(TestCase):
+
+    def test_filter(self):
+        library = Library()
+
+        @library.filter
+        def func():
+            return ''
+
+        self.assertEqual(library.filters['func'], func)
+
+    def test_filter_parens(self):
+        library = Library()
+
+        @library.filter()
+        def func():
+            return ''
+
+        self.assertEqual(library.filters['func'], func)
+
+    def test_filter_name_arg(self):
+        library = Library()
+
+        @library.filter('name')
+        def func():
+            return ''
+
+        self.assertEqual(library.filters['name'], func)
+
+    def test_filter_name_kwarg(self):
+        library = Library()
+
+        @library.filter(name='name')
+        def func():
+            return ''
+
+        self.assertEqual(library.filters['name'], func)
+
+    def test_filter_call(self):
+        library = Library()
+
+        def func():
+            return ''
+        library.filter('name', func)
+
+        self.assertEqual(library.filters['name'], func)
+
+    def test_filter_invalid(self):
+        library = Library()
+        msg = "Unsupported arguments to Library.filter: (None, '')"
+        with self.assertRaisesMessage(ValueError, msg):
+            library.filter(None, '')
+
+
+class InclusionTagRegistrationTests(TestCase):
+
+    def test_inclusion_tag(self):
+        library = Library()
+
+        @library.inclusion_tag('template.html')
+        def func():
+            return ''
+
+        self.assertTrue('func' in library.tags)
+
+    def test_inclusion_tag_name(self):
+        library = Library()
+
+        @library.inclusion_tag('template.html', name='name')
+        def func():
+            return ''
+
+        self.assertTrue('name' in library.tags)
+
+
+class SimpleTagRegistrationTests(TestCase):
+
+    def test_simple_tag(self):
+        library = Library()
+
+        @library.simple_tag
+        def func():
+            return ''
+
+        self.assertTrue('func' in library.tags)
+
+    def test_simple_tag_parens(self):
+        library = Library()
+
+        @library.simple_tag()
+        def func():
+            return ''
+
+        self.assertTrue('func' in library.tags)
+
+    def test_simple_tag_name_kwarg(self):
+        library = Library()
+
+        @library.simple_tag(name='name')
+        def func():
+            return ''
+
+        self.assertTrue('name' in library.tags)
+
+    def test_simple_tag_invalid(self):
+        library = Library()
+        msg = "Invalid arguments provided to simple_tag"
+        with self.assertRaisesMessage(ValueError, msg):
+            library.simple_tag('invalid')
+
+
+class TagRegistrationTests(TestCase):
+
+    def test_tag(self):
+        library = Library()
+
+        @library.tag
+        def func(parser, token):
+            return Node()
+        self.assertEqual(library.tags['func'], func)
```

Comment:
in most places you have a blank line after the function before the assert, but not all